File: iRINGTools.ESBServices/iRINGTools.Common/src/org/iringtools/security/LdapAuthorizationProvider.java
package org.iringtools.security;

import java.util.Map;
import java.util.Map.Entry;
import java.util.Properties;

import javax.naming.NamingEnumeration;
import javax.naming.NamingException;
import javax.naming.directory.DirContext;
import javax.naming.directory.InitialDirContext;
import javax.naming.directory.SearchControls;

import org.apache.log4j.Logger;
import org.iringtools.utility.EncryptionException;
import org.iringtools.utility.EncryptionUtils;

public class LdapAuthorizationProvider implements AuthorizationProvider
{
  private static final Logger logger = Logger.getLogger(LdapAuthorizationProvider.class);
  
  private static final String BASE_DN = "ou=iringtools,dc=iringug,dc=org";
  private static final String USERID_KEY = "EmailAddress";
  
  private DirContext dctx;
  private String authorizedGroup;

  public void init(Properties properties) throws NamingException
  {
    String server = properties.getProperty("server");
    String portNumber = properties.getProperty("portNumber");
    String userName = properties.getProperty("userName");
    String password = properties.getProperty("password");
    
    authorizedGroup = properties.getProperty("authorizedGroup");
    
    try
    {
      password = EncryptionUtils.decrypt(password);
    }
    catch (EncryptionException e)
    {
      logger.error("Error decrypting ldap password [" + password + "]");
    }

    Properties ldapConfig = new Properties();
    ldapConfig.put("java.naming.factory.initial", "com.sun.jndi.ldap.LdapCtxFactory");
    ldapConfig.put("java.naming.provider.url", "ldap://" + server + ":" + portNumber);
    ldapConfig.put("java.naming.security.authentication", "simple");
    ldapConfig.put("java.naming.security.principal", userName);
    ldapConfig.put("java.naming.security.credentials", password);
    
    dctx = new InitialDirContext(ldapConfig);
  }
  
  public boolean isAuthorized(Map<String, String> claims)
  {
    String userId = getUserId(claims);

    if (userId != null && dctx != null)
    {
      String groupDN = "cn=" + authorizedGroup + ",cn=groups," + BASE_DN;
      String qualUserId = "uid=" + userId + ",cn=users," + BASE_DN;
      String filter = "(member=" + qualUserId + ")";

      SearchControls constraints = new SearchControls();
      constraints.setSearchScope(SearchControls.SUBTREE_SCOPE);

      try
      {
        NamingEnumeration<?> results = dctx.search(groupDN, filter, constraints);

        if (results != null && results.hasMore())
        {
          return true;
        }
      }
      catch (NamingException ex)
      {
        logger.error("User [" + userId + "] not authorized.");
      }
    }

    return false;
  }

  private String getUserId(Map<String, String> claims)
  {
    for (Entry<String, String> entry : claims.entrySet())
    {
      if (entry.getKey().equalsIgnoreCase(USERID_KEY))
      {
        return entry.getValue();
      }
    }

    return null;
  }
}

File: Libraries/iRINGLibrary/Desktop/LdapAuthorizationProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using log4net;
using System.DirectoryServices.Protocols;
using System.Net;
using org.iringtools.utility;
using System.Collections;

namespace org.iringtools.adapter.security
{
  public class LdapAuthorizationProvider : IAuthorizationLayer
  {
    private static readonly ILog _logger = LogManager.GetLogger(typeof(LdapAuthorizationProvider));

    private const string BASE_DN = "o=iringtools,dc=iringug,dc=org";
    private const string USERID_KEY = "emailaddress";
    
    private LdapConnection ldapConnection;
    private string authorizedGroup;

    public void Init(Properties properties)
    {
      string server = properties["server"];
      int portNumber = Int32.Parse(properties["portNumber"]);
      string userName = properties["userName"];
      string password = EncryptionUtility.Decrypt(properties["password"]);

      LdapDirectoryIdentifier ldapIdentifier = new LdapDirectoryIdentifier(server, portNumber, true, false);
      ldapConnection = new LdapConnection(ldapIdentifier);
      ldapConnection.Credential = new NetworkCredential(userName, password);
      ldapConnection.AuthType = AuthType.Basic;
      ldapConnection.Bind();

      authorizedGroup = properties["authorizedGroup"];
    }

    public bool IsAuthorized(IDictionary claims)
    {
      string userId = GetUserId((IDictionary<string, string>)claims);

      if (userId != null && ldapConnection != null)
      {
        string groupDN = "cn=" + authorizedGroup + ",ou=groups," + BASE_DN;
        string qualUserId = "uid=" + userId + ",ou=users," + BASE_DN;
        string filter = "(member=" + qualUserId + ")";

        SearchRequest request = new SearchRequest
        {
          DistinguishedName = groupDN,
          Filter = filter,
          Scope = System.DirectoryServices.Protocols.SearchScope.Subtree,
        };

        SearchResponse response = (SearchResponse)ldapConnection.SendRequest(request);
        UTF8Encoding utf8 = new UTF8Encoding(false, true);

        return (response.Entries.Count > 0);
      }

      return false;
    }

    private string GetUserId(IDictionary<string, string> claims)
    {
      foreach (var pair in claims)
      {
        if (pair.Key.ToLower() == USERID_KEY)
        {
          return pair.Value;
        }
      }

      return null;
    }
  }
}
